implementations/micrometer-registry-cloudwatch/src/main/java/io/micrometer/cloudwatch/MetricDatumPartition.java

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright 2017 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.cloudwatch;
+
+import com.amazonaws.services.cloudwatch.model.MetricDatum;
+import com.google.common.math.IntMath;
+
+import java.math.RoundingMode;
+import java.util.AbstractList;
+import java.util.List;
+
+/**
+ * Modified from {@link io.micrometer.core.instrument.util.MeterPartition}
+ */
+public class MetricDatumPartition extends AbstractList<List<MetricDatum>> {
+    private final List<MetricDatum> list;
+    private final int size;
+
+    public MetricDatumPartition(List<MetricDatum> metricData, int size) {
+        this.list = metricData;
+        this.size = size;
+    }
+
+    @Override
+    public List<MetricDatum> get(int index) {
+        int start = index * size;
+        int end = Math.min(start + size, list.size());
+        return list.subList(start, end);
+    }
+
+    @Override
+    public int size() {
+        return IntMath.divide(list.size(), size, RoundingMode.CEILING);
```

Comment:
While Guava is available as an optional dep, you might want to do this long hand so that it isn't a hard dependency for using AWS cloudwatch.

Reply:
I started with MeterRegistry and CumulativeCounter but then ClouWatch counter metrics doesn't behave as I would expect to.

With CloudWatch newly sent metric is being aggregated into an existing metric so there is no access to original metrics but only to aggregated form (Avg, Min, Max, Sum, Sample count, percentiles), with default one minute resolution.


eg. test: every 5 seconds I increase counter by 10,
expectation: 12 * 10 = 120 avg outcome for every minute,


with cumulative counter:

time		00	05	10	15	20	25	30	35	40	45	50	55	
counter 	10	20	30	40	50	60	70	80	90	100	110	120

e.g. sampling: 06, 26, 46
avg: (20 + 60 + 100) / 3 = 60
min: 20
max: 100
sum: 180

none of these is of any usage and all of them just go up and up until everything is messeed up even more with server restart.


with step counter:

time		00	05	10	15	20	25	30	35	40	45	50	55	
counter 	10	20	10	20	30	40	10	20	30	40	10	20

e.g. sampling: 06, 26, 46
avg: (20 + 40 + 40) / 3 = 33.3 (40 every minute after first)
min: 20 (40 every minute after first)
max: 40
sum: 100 (120 every minute after first)

This can be generalized to other sampling periods and and non regular metrics sending.

similar things can be reason about Timer and StepTimer.

stepCounter Sum:
<img width="772" alt="stepcounter" src="https://user-images.githubusercontent.com/5848865/33214114-cef95fc0-d12a-11e7-8988-9911c1bf289a.png">

cumulativeCounter Avg
<img width="805" alt="cumulativecounter" src="https://user-images.githubusercontent.com/5848865/33214118-d2ca6482-d12a-11e7-9d27-034048dbee31.png">

Reply:
But this one is done long hand and doesn't use guava :)

Reply:
I guess I'm confused by the import: `com.google.common.math.IntMath`

Reply:
And the usage of it. It's clearly using Google guava

Reply:
sorry, somehow I missed it! I removed this dependency in new commit